airflow/models/renderedtifields.py

```diff
@@ -0,0 +1,172 @@
+    """
+    Save Rendered Template Fields
+    """
+
+    __tablename__ = "rendered_task_instance_fields"
+
+    dag_id = Column(String(ID_LEN), primary_key=True)
+    task_id = Column(String(ID_LEN), primary_key=True)
+    execution_date = Column(UtcDateTime, primary_key=True)
+    rendered_fields = Column(JSON, nullable=False)
+
+    def __init__(self, ti: TaskInstance, render_templates=True):
+        self.dag_id = ti.dag_id
+        self.task_id = ti.task_id
+        self.task = ti.task
+        self.execution_date = ti.execution_date
+        self.ti = ti
+        if render_templates:
+            ti.render_templates()
+        self.rendered_fields = {
+            field: serialize_template_field(
+                getattr(self.task, field)
+            ) for field in self.task.template_fields
+        }
+
+    def __repr__(self):
+        return f"<{self.__class__.__name__}: {self.dag_id}.{self.task_id} {self.execution_date}"
+
+    @classmethod
+    @provide_session
+    def get_templated_fields(cls, ti: TaskInstance, session: Session = None) -> Optional[dict]:
+        """
+        Get templated field for a TaskInstance from the RenderedTaskInstanceFields
+        table.
+
+        :param ti: Task Instance
+        :param session: SqlAlchemy Session
+        :return: Rendered Templated TI field
+        """
+        result = session.query(cls.rendered_fields).filter(
+            cls.dag_id == ti.dag_id,
+            cls.task_id == ti.task_id,
+            cls.execution_date == ti.execution_date
+        ).first()
+
+        if result:
+            rendered_fields = result.rendered_fields
+            if isinstance(rendered_fields, str):
+                rendered_fields = json.loads(rendered_fields)
+            return rendered_fields
+        else:
+            return None
+
+    @classmethod
+    @provide_session
+    def has_templated_fields(cls, ti: TaskInstance, session: Session = None) -> bool:
+        """Checks templated field exist for this ti.
+
+        :param ti: Task Instance
+        :param session: SqlAlchemy Session
+        """
+        # We cant use q.exists() because some databases such as SQL Server don’t allow
+        # an EXISTS expression to be present in the columns clause of a SELECT.
+        # Details: https://docs.sqlalchemy.org/en/13/orm/query.html#sqlalchemy.orm.query.Query.exists
+        return session.query(exists().where(
+            and_(cls.dag_id == ti.dag_id,
+                 cls.task_id == ti.task_id,
+                 cls.execution_date == ti.execution_date)
+        )).scalar()
+
+    @provide_session
+    def write(self, overwrite: bool = False, session: Session = None):
+        """Write instance to database
+
+        :param overwrite: Override if record exists in DB
+        :param session: SqlAlchemy Session
+        """
+        if not self.has_templated_fields(self.ti, session):
+            session.add(self)
+        else:
+            if overwrite:
```

Comment:
We do nothing of this is false, even if the row doesn't yet exist

Reply:
I wonder if this entire function could become `session.merge` - which would do a upsert for us.

Reply:
Yes, I agree. Will update

Reply:
Updated